bundles/org.openhab.binding.alarmdecoder/src/main/java/org/openhab/binding/alarmdecoder/internal/handler/SerialBridgeHandler.java

```diff
@@ -0,0 +1,163 @@
+import org.eclipse.smarthome.core.thing.ThingStatusDetail;
+import org.eclipse.smarthome.io.transport.serial.PortInUseException;
+import org.eclipse.smarthome.io.transport.serial.SerialPort;
+import org.eclipse.smarthome.io.transport.serial.SerialPortIdentifier;
+import org.eclipse.smarthome.io.transport.serial.SerialPortManager;
+import org.eclipse.smarthome.io.transport.serial.UnsupportedCommOperationException;
+import org.openhab.binding.alarmdecoder.internal.config.SerialBridgeConfig;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Handler responsible for communicating via a serial port with the Nu Tech Alarm Decoder device.
+ * Based on code from the original OH1 alarmdecoder binding. Some OHC serial transport code taken from the Zigbee
+ * binding.
+ *
+ * @author Bernd Pfrommer - Initial contribution (OH1 version)
+ * @author Bob Adair - Re-factored into OH2 binding and rewrote to use OHC serial transport.
+ */
+@NonNullByDefault
+public class SerialBridgeHandler extends ADBridgeHandler {
+
+    private final Logger logger = LoggerFactory.getLogger(SerialBridgeHandler.class);
+
+    private @NonNullByDefault({}) SerialBridgeConfig config;
+    private final SerialPortManager serialPortManager;
+
+    /** name of serial device */
+    private String serialDeviceName = "";
+    private @NonNullByDefault({}) SerialPort serialPort = null;
+    private int serialPortSpeed = 115200;
+
+    public SerialBridgeHandler(Bridge bridge, SerialPortManager serialPortManager) {
+        super(bridge);
+        this.serialPortManager = serialPortManager;
+    }
+
+    @Override
+    public void initialize() {
+        logger.debug("Initializing serial bridge handler");
+        config = getConfigAs(SerialBridgeConfig.class);
+        discovery = config.discovery;
+
+        if (config.serialPort == null) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "bridge configuration missing");
+            return;
+        } else {
+            serialDeviceName = config.serialPort;
+        }
+
+        if (config.bitrate > 0) {
+            serialPortSpeed = config.bitrate;
+        }
+
+        connect();
+
+        logger.trace("Finished initializing serial bridge handler");
+    }
+
+    @Override
+    protected synchronized void connect() {
+        disconnect(); // make sure we are disconnected
+        try {
+            if (!serialDeviceName.isEmpty()) {
+                // Exit if no identifiers exist to work around possible bug
+                Stream<SerialPortIdentifier> serialPortIdentifiers = serialPortManager.getIdentifiers();
+                if (!serialPortIdentifiers.findAny().isPresent()) {
+                    logger.debug("No serial communication ports found. Cannot connect to [{}]", serialDeviceName);
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "No serial ports found");
+                    return;
+                }
```

Comment:
Maybe. I moved it, but I'm hoping that we can get rid of it in 3.0 since the new serial library update. I added a comment to that effect. See openhab/openhab-core/#1426